File: src/tools/characters.ts
import { v4 as uuidv4 } from "uuid";
import { getDatabase } from "../db/connection.js";
import type { Character, VoiceConfig } from "../db/schema.js";

export interface AddCharacterParams {
  name: string;
  description?: string;
  is_narrator?: boolean;
}

export interface UpdateCharacterParams {
  id: string;
  name?: string;
  description?: string;
  is_narrator?: boolean;
}

export interface SetVoiceParams {
  character_id: string;
  provider: string;
  voice_id: string;
  settings?: Record<string, unknown>;
}

/**
 * Add a new character to the project
 */
export function addCharacter(params: AddCharacterParams): Character {
  const db = getDatabase();
  const { name, description, is_narrator = false } = params;

  // Check for duplicate name
  const existing = db
    .prepare(`SELECT id FROM characters WHERE name = ?`)
    .get(name);
  if (existing) {
    throw new Error(`Character with name "${name}" already exists`);
  }

  const id = uuidv4();
  const now = new Date().toISOString();

  const stmt = db.prepare(`
    INSERT INTO characters (id, name, description, voice_config, is_narrator, created_at)
    VALUES (?, ?, ?, NULL, ?, ?)
  `);

  stmt.run(id, name, description ?? null, is_narrator ? 1 : 0, now);

  return {
    id,
    name,
    description: description ?? null,
    voice_config: null,
    is_narrator,
    created_at: now,
  };
}

/**
 * List all characters in the project
 */
export function listCharacters(): Character[] {
  const db = getDatabase();

  const stmt = db.prepare(`
    SELECT * FROM characters ORDER BY is_narrator DESC, name ASC
  `);

  const rows = stmt.all() as Array<Record<string, unknown>>;

  return rows.map((row) => ({
    id: row.id as string,
    name: row.name as string,
    description: row.description as string | null,
    voice_config: row.voice_config as string | null,
    is_narrator: Boolean(row.is_narrator),
    created_at: row.created_at as string,
  }));
}

/**
 * Get a character by ID
 */
export function getCharacter(id: string): Character | null {
  const db = getDatabase();

  const stmt = db.prepare(`SELECT * FROM characters WHERE id = ?`);
  const row = stmt.get(id) as Record<string, unknown> | undefined;

  if (!row) return null;

  return {
    id: row.id as string,
    name: row.name as string,
    description: row.description as string | null,
    voice_config: row.voice_config as string | null,
    is_narrator: Boolean(row.is_narrator),
    created_at: row.created_at as string,
  };
}

/**
 * Update a character
 */
export function updateCharacter(params: UpdateCharacterParams): Character {
  const db = getDatabase();
  const { id, name, description, is_narrator } = params;

  // Verify character exists
  const existing = getCharacter(id);
  if (!existing) {
    throw new Error(`Character not found: ${id}`);
  }

  // Check for name conflict if renaming
  if (name && name !== existing.name) {
    const nameConflict = db
      .prepare(`SELECT id FROM characters WHERE name = ? AND id != ?`)
      .get(name, id);
    if (nameConflict) {
      throw new Error(`Character with name "${name}" already exists`);
    }
  }

  // Build update query
  const updates: string[] = [];
  const values: (string | number | null)[] = [];

  if (name !== undefined) {
    updates.push("name = ?");
    values.push(name);
  }
  if (description !== undefined) {
    updates.push("description = ?");
    values.push(description);
  }
  if (is_narrator !== undefined) {
    updates.push("is_narrator = ?");
    values.push(is_narrator ? 1 : 0);
  }

  if (updates.length === 0) {
    return existing;
  }

  values.push(id);
  const stmt = db.prepare(
    `UPDATE characters SET ${updates.join(", ")} WHERE id = ?`
  );
  stmt.run(...values);

  return getCharacter(id)!;
}

/**
 * Delete a character
 */
export function deleteCharacter(id: string): void {
  const db = getDatabase();

  // Verify character exists
  const existing = getCharacter(id);
  if (!existing) {
    throw new Error(`Character not found: ${id}`);
  }

  const stmt = db.prepare(`DELETE FROM characters WHERE id = ?`);
  stmt.run(id);
}

/**
 * Set voice configuration for a character
 */
export function setVoice(params: SetVoiceParams): Character {
  const db = getDatabase();
  const { character_id, provider, voice_id, settings } = params;

  // Verify character exists
  const existing = getCharacter(character_id);
  if (!existing) {
    throw new Error(`Character not found: ${character_id}`);
  }

  const voiceConfig: VoiceConfig = {
    provider,
    voice_id,
    settings,
  };

  const stmt = db.prepare(`UPDATE characters SET voice_config = ? WHERE id = ?`);
  stmt.run(JSON.stringify(voiceConfig), character_id);

  return getCharacter(character_id)!;
}

/**
 * Clear voice configuration for a character
 */
export function clearVoice(character_id: string): Character {
  const db = getDatabase();

  // Verify character exists
  const existing = getCharacter(character_id);
  if (!existing) {
    throw new Error(`Character not found: ${character_id}`);
  }

  const stmt = db.prepare(`UPDATE characters SET voice_config = NULL WHERE id = ?`);
  stmt.run(character_id);

  return getCharacter(character_id)!;
}

/**
 * Get characters with their line counts
 */
export function getCharactersWithStats(): Array<
  Character & { segment_count: number; total_text_length: number }
> {
  const db = getDatabase();

  const stmt = db.prepare(`
    SELECT
      c.*,
      COUNT(s.id) as segment_count,
      COALESCE(SUM(LENGTH(s.text_content)), 0) as total_text_length
    FROM characters c
    LEFT JOIN segments s ON s.character_id = c.id
    GROUP BY c.id
    ORDER BY c.is_narrator DESC, c.name ASC
  `);

  const rows = stmt.all() as Array<Record<string, unknown>>;

  return rows.map((row) => ({
    id: row.id as string,
    name: row.name as string,
    description: row.description as string | null,
    voice_config: row.voice_config as string | null,
    is_narrator: Boolean(row.is_narrator),
    created_at: row.created_at as string,
    segment_count: row.segment_count as number,
    total_text_length: row.total_text_length as number,
  }));
}

File: audiobook_mcp/tools/characters.py
"""Character management tools."""

import json
import uuid
from datetime import datetime
from typing import Optional, Any
from dataclasses import dataclass

from ..db.connection import get_database
from ..db.schema import Character, VoiceConfig


@dataclass
class CharacterWithStats:
    id: str
    name: str
    description: Optional[str]
    voice_config: Optional[str]
    is_narrator: bool
    created_at: str
    segment_count: int
    total_text_length: int


def add_character(
    name: str,
    description: Optional[str] = None,
    is_narrator: bool = False,
) -> Character:
    """Add a new character to the project."""
    db = get_database()
    cursor = db.cursor()

    # Check for duplicate name
    cursor.execute("SELECT id FROM characters WHERE name = ?", (name,))
    if cursor.fetchone():
        raise ValueError(f'Character with name "{name}" already exists')

    character_id = str(uuid.uuid4())
    now = datetime.utcnow().isoformat() + "Z"

    cursor.execute(
        """
        INSERT INTO characters (id, name, description, voice_config, is_narrator, created_at)
        VALUES (?, ?, ?, NULL, ?, ?)
        """,
        (character_id, name, description, 1 if is_narrator else 0, now),
    )
    db.commit()

    return Character(
        id=character_id,
        name=name,
        description=description,
        voice_config=None,
        is_narrator=is_narrator,
        created_at=now,
    )


def list_characters() -> list[Character]:
    """List all characters in the project."""
    db = get_database()
    cursor = db.cursor()

    cursor.execute("SELECT * FROM characters ORDER BY is_narrator DESC, name ASC")
    rows = cursor.fetchall()

    return [
        Character(
            id=row["id"],
            name=row["name"],
            description=row["description"],
            voice_config=row["voice_config"],
            is_narrator=bool(row["is_narrator"]),
            created_at=row["created_at"],
        )
        for row in rows
    ]


def get_character(character_id: str) -> Optional[Character]:
    """Get a character by ID."""
    db = get_database()
    cursor = db.cursor()

    cursor.execute("SELECT * FROM characters WHERE id = ?", (character_id,))
    row = cursor.fetchone()

    if not row:
        return None

    return Character(
        id=row["id"],
        name=row["name"],
        description=row["description"],
        voice_config=row["voice_config"],
        is_narrator=bool(row["is_narrator"]),
        created_at=row["created_at"],
    )


def update_character(
    character_id: str,
    name: Optional[str] = None,
    description: Optional[str] = None,
    is_narrator: Optional[bool] = None,
) -> Character:
    """Update a character."""
    db = get_database()
    cursor = db.cursor()

    # Verify character exists
    existing = get_character(character_id)
    if not existing:
        raise ValueError(f"Character not found: {character_id}")

    # Check for name conflict if renaming
    if name and name != existing.name:
        cursor.execute(
            "SELECT id FROM characters WHERE name = ? AND id != ?",
            (name, character_id),
        )
        if cursor.fetchone():
            raise ValueError(f'Character with name "{name}" already exists')

    # Build update query
    updates = []
    values: list = []

    if name is not None:
        updates.append("name = ?")
        values.append(name)
    if description is not None:
        updates.append("description = ?")
        values.append(description)
    if is_narrator is not None:
        updates.append("is_narrator = ?")
        values.append(1 if is_narrator else 0)

    if not updates:
        return existing

    values.append(character_id)
    cursor.execute(
        f"UPDATE characters SET {', '.join(updates)} WHERE id = ?",
        values,
    )
    db.commit()

    return get_character(character_id)  # type: ignore


def delete_character(character_id: str) -> None:
    """Delete a character."""
    db = get_database()
    cursor = db.cursor()

    # Verify character exists
    existing = get_character(character_id)
    if not existing:
        raise ValueError(f"Character not found: {character_id}")

    cursor.execute("DELETE FROM characters WHERE id = ?", (character_id,))
    db.commit()


def set_voice(
    character_id: str,
    provider: str,
    voice_id: str,
    settings: Optional[dict[str, Any]] = None,
) -> Character:
    """Set voice configuration for a character."""
    db = get_database()
    cursor = db.cursor()

    # Verify character exists
    existing = get_character(character_id)
    if not existing:
        raise ValueError(f"Character not found: {character_id}")

    voice_config = VoiceConfig(provider=provider, voice_id=voice_id, settings=settings)
    voice_config_json = json.dumps({
        "provider": voice_config.provider,
        "voice_id": voice_config.voice_id,
        "settings": voice_config.settings,
    })

    cursor.execute(
        "UPDATE characters SET voice_config = ? WHERE id = ?",
        (voice_config_json, character_id),
    )
    db.commit()

    return get_character(character_id)  # type: ignore


def clear_voice(character_id: str) -> Character:
    """Clear voice configuration for a character."""
    db = get_database()
    cursor = db.cursor()

    # Verify character exists
    existing = get_character(character_id)
    if not existing:
        raise ValueError(f"Character not found: {character_id}")

    cursor.execute(
        "UPDATE characters SET voice_config = NULL WHERE id = ?",
        (character_id,),
    )
    db.commit()

    return get_character(character_id)  # type: ignore


def get_characters_with_stats() -> list[CharacterWithStats]:
    """Get characters with their line counts."""
    db = get_database()
    cursor = db.cursor()

    cursor.execute("""
        SELECT
            c.*,
            COUNT(s.id) as segment_count,
            COALESCE(SUM(LENGTH(s.text_content)), 0) as total_text_length
        FROM characters c
        LEFT JOIN segments s ON s.character_id = c.id
        GROUP BY c.id
        ORDER BY c.is_narrator DESC, c.name ASC
    """)
    rows = cursor.fetchall()

    return [
        CharacterWithStats(
            id=row["id"],
            name=row["name"],
            description=row["description"],
            voice_config=row["voice_config"],
            is_narrator=bool(row["is_narrator"]),
            created_at=row["created_at"],
            segment_count=row["segment_count"],
            total_text_length=row["total_text_length"],
        )
        for row in rows
    ]
